src/main/java/org/jenkinsci/plugins/workflow/steps/CatchErrorStep.java

```diff
@@ -53,59 +125,98 @@ public final class CatchErrorStep extends Step {
+            return r;
+        }
+
+        public ListBoxModel doFillStageResultItems() {
+            return doFillBuildResultItems();
         }
 
     }
 
     public static final class Execution extends AbstractStepExecutionImpl {
+        @SuppressFBWarnings(value="SE_TRANSIENT_FIELD_NOT_RESTORED", justification="Only used at startup, serialized in Callback")
+        private transient final CatchExecutionOptions options;
 
-        Execution(StepContext context) {
+        Execution(StepContext context, CatchExecutionOptions options) {
             super(context);
+            this.options = options;
         }
 
         @Override public boolean start() throws Exception {
             StepContext context = getContext();
             context.newBodyInvoker()
-                    .withCallback(new Callback())
+                    .withCallback(new Callback(options))
                     .start();
             return false;
         }
 
         @Override public void onResume() {}
 
         private static final class Callback extends BodyExecutionCallback {
+            private static final long serialVersionUID = -5448044884830236797L;
+            private CatchExecutionOptions options;
+
+            public Callback(CatchExecutionOptions options) {
+                this.options = options;
+            }
+
+            public Object readResolve() {
+                if (options == null) {
+                    options = DEFAULT_OPTIONS;
+                }
+                return this;
+            }
 
             @Override public void onSuccess(StepContext context, Object result) {
                 context.onSuccess(null); // we do not pass up a result, since onFailure cannot
             }
 
             @Override public void onFailure(StepContext context, Throwable t) {
                 try {
+                    if (!options.isCatchInterruptions() && t instanceof FlowInterruptedException) {
+                        context.onFailure(t);
+                        return;
+                    }
                     TaskListener listener = context.get(TaskListener.class);
-                    Result r = Result.FAILURE;
+                    String message = options.getMessage();
+                    if (message != null) {
+                        listener.error(message);
+                    }
+                    Result buildResult = options.getBuildResultOnError();
+                    Result stepResult = options.getStepResultOnError();
                     if (t instanceof AbortException) {
                         listener.error(t.getMessage());
                     } else if (t instanceof FlowInterruptedException) {
                         FlowInterruptedException fie = (FlowInterruptedException) t;
                         fie.handle(context.get(Run.class), listener);
-                        r = fie.getResult();
+                        buildResult = fie.getResult();
+                        stepResult = fie.getResult();
```

Comment:
I am confused by this logic. If I kept `catchInterruptions` on, and set `buildResult` and `stageResult` to null, I would expect that anything that happens inside the `catchError` step could not fail the build. Yet, apparently, a `FlowInterruptedException` can still fail the build. Am I missing something?

Reply:
I do not remember if this was intentional or not. My only guess it that the idea was to preserve the old behavior where `FlowInterruptedException.result` was always used to set the build result. If it does not break existing tests then I am not opposed to deleting these lines, or perhaps it could make sense to continue to use `FlowInterruptedException.result` but only when it `Result.isBetterThan` `getBuildResultOnError`/`getStepResultOnError`.

Reply:
Not sure. In the meantime, the workaround for the record is to just use `try`…`catch`.